tool/planet/environment.go

```diff
@@ -0,0 +1,100 @@
+package main
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/planet/lib/box"
+	"github.com/gravitational/planet/lib/constants"
+	"github.com/gravitational/planet/lib/utils"
+
+	etcdconf "github.com/gravitational/coordinate/config"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+func updateEnvironment(kvs map[string]string) error {
+	log.WithField("kvs", kvs).Info("Update environment.")
+
+	env, err := box.ReadEnvironment(ContainerEnvironmentFile)
+	if err != nil {
+		return trace.Wrap(err, "failed to read cluster environment file")
+	}
+
+	for k, v := range kvs {
+		env.Upsert(k, v)
+	}
+
+	err = utils.SafeWriteFile(ContainerEnvironmentFile, env, constants.SharedReadMask)
+	if err != nil {
+		return trace.Wrap(err, "failed to write cluster environment file")
+	}
+
+	leaderAddr, err := getCurrentLeader(env)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	isLeader := env.Get(EnvPublicIP) == leaderAddr
+	services := append(commonServices, kubeServices...)
+	if isLeader {
+		services = append(services, kubeMasterServices...)
+	}
+	return restartServices(context.Background(), services)
```

Comment:
Have you considered taking even more granular approach i.e. have separate phases for restart of each service? Then if a certain service failed to restart, it will be easy to see what failed and rollback. And this function would only update the container-environment file. With this "batch restart" approach, wouldn't it be harder to rollback should something goes wrong?

Reply:
I have and I don't think it would work. The first issue is that I cannot seem to figure out how to handle rollbacks in proper order given separate steps for services:
i.e. normal order:
  1. update container environment file
  1. restart service 1
  1. ...
  1. restart service n

For rollback, I'd still need step 1 before the rest - it does not work like that now as rollback steps are executed last to first. Besides, since requirements are merely recommendations and are only validated for forward execution, the rollbacks do not take them into account. Not sure how much time it would cost to change that as well w/o breaking the rest of operations.